### PR TITLE
Feature / Updated transformer weight before transformer ranking

### PR DESCRIPTION
**Pull request by Jerry-Jinfeng-Guo**
This PR implements the updated transformer ranking in https://github.com/PowerGridModel/power-grid-model/issues/662

* [x] Unregulated transformer edges get weight `0` instead of `1`
* [x] Update unit tests
  * [x] Add unit tests with un-regulated transformers
  * [x] Update existing tests based on new correct results
  * [x] Update docs

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 1]
```diff
@@ -113,11 +113,14 @@ inline void process_trafo3w_edge(main_core::main_model_state_c auto const& state
             auto const& non_tap_side_node = tap_at_first_side ? to_node : from_node;
             // add regulated idx only when the first side node is tap side node.
             // This is done to add only one directional edge with regulated idx.
-            Idx2D const regulated_idx = from_node == tap_side_node ? unregulated_idx : trafo3w_idx;
```
Comment (figueroa1395) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:116]: Just for context, what does `trafo3w_idx` stand for? Transformer three winding index?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 2]
```diff
@@ -113,11 +113,14 @@ inline void process_trafo3w_edge(main_core::main_model_state_c auto const& state
...
-            add_to_edge(state, edges, edge_props, tap_side_node, non_tap_side_node, {regulated_idx, 1});
+            if (from_node == tap_side_node) {
+                add_to_edge(state, edges, edge_props, tap_side_node, non_tap_side_node, {unregulated_idx, 0});
+            } else {
+                add_to_edge(state, edges, edge_props, tap_side_node, non_tap_side_node, {trafo3w_idx, 1});
```
Comment (figueroa1395) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:119]: Why this change to if-else? Is it because unlike the `unregulated_idx`, the `trafo3w_idx = 1` for the ranking weight (didn't change)?

**Review by figueroa1395 — COMMENTED** [review]
Looks good to me, but just to double check, some Python tests don't need to be adjusted (besides the validation part) in the same way as the Cpp one?

Note: VS-Code was stroking on me when I tried to review from there, so duplicated comments may randomly appear at some point. If I catch them first I will delete, if someone else does, please delete or resolve.

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 1 above)
Comment (nitbharambe) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:116]: Yes

**Review by mgovers — COMMENTED** [review]
LGTM. not yet approving due to remaining open comment, but otherwise OK

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 2 above)
Comment (Jerry-Jinfeng-Guo) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp]: Addressed. For readability, a variable is used.

**Review by mgovers — APPROVED** [review]
